### PR TITLE
Add TLS support to Torii

### PR DESCRIPTION
**Pull request by ionagamed**
Signed-off-by: Leonid Lygin <ionagamed@gmail.com>

### Description of the Change

This adds optional TLS encryption for torii on a separate port

### Benefits

Improved security of client-node communication

### Possible Drawbacks 

None

### CODE REVIEW

File: docs/source/guides/sec-install.rst [hunk 1]
```diff
@@ -21,12 +21,15 @@ Create an encrypted backup of private keys before deleting them and limit the ac
...
 Network configuration
 ^^^^^^^^^^^^^^^^^^^^^
-Iroha listens on ports 50051 and 10001.
+Iroha listens on ports 50051 and 10001, and optionally 50052, if TLS is enabled.
 Firewall settings must allow incoming/outgoing connections to/from these ports.
 If possible, disable or remove any other network services with listening ports (FTP, DNS, LDAP, SMB, DHCP, NFS, SNMP, etc).
 Ideally, Iroha should be as much isolated as possible in terms of networking.
 
-Currently, there is no traffic encryption in Iroha, we strongly recommend using VPN or Calico for setting up Docker overlay network, i.e. any mechanism that allows encrypting communication between peers.
+You can enable TLS on torii ports if you would like to encrypt client-peer
```
Comment (igor-egorov) [inline docs/source/guides/sec-install.rst]:
```suggestion
You can enable TLS on torii ports if you would like to encrypt client-peer
```
(trailing whitespace)

File: docs/source/guides/torii-tls.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+Keys must be presented in PEM format. To generate them you can use ``openssl``:
+
+```bash
+$ openssl genpkey -algorithm <desired algorithm> -out server.key
+$ openssl req -new -key server.key -x509 -out server.crt
+```
+
+If you need to use plain IP addresses to connect to the node, you need to
+specify subjectAltName in your server certificate, for that you need to add
+a ``subjectAltName`` directive to ``v3_ca`` section of your openssl config
```
Comment (igor-egorov) [inline docs/source/guides/torii-tls.rst]: Please install some plugin to your editor that shows or automatically removes trailing whitespaces.
```suggestion
You can enable TLS on torii ports if you would like to encrypt client-peer
```

File: docs/source/guides/torii-tls.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+```bash
+$ openssl genpkey -algorithm <desired algorithm> -out server.key
+$ openssl req -new -key server.key -x509 -out server.crt
+```
+
+If you need to use plain IP addresses to connect to the node, you need to
+specify subjectAltName in your server certificate, for that you need to add
+a ``subjectAltName`` directive to ``v3_ca`` section of your openssl config 
+before generating the certificate.
```
Comment (igor-egorov) [inline docs/source/guides/torii-tls.rst]: trailing whitespace
```suggestion
before generating the certificate.
```

File: docs/source/guides/torii-tls.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+```conf
+[ v3_ca ]
+subjectAltName=IP:12.34.56.78
+...
+```
+
+Fields in the certificate don't really matter except for the Common Name (CN),
+it would be checked against the client's hostname, and TLS handshake will fail
+if they do not match (e.g. if you connect to example.com:50051, then irohad at
```
Comment (igor-egorov) [inline docs/source/guides/torii-tls.rst]:
```suggestion
if they do not match (e.g. if you connect to example.com:50051, then irohad at
```
trailing whitespace

File: docs/source/guides/torii-tls.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+To configure iroha to use your keys, you need to modify ``torii_tls_port`` and
+``torii_tls_keypair`` config parameters.
+
+``torii_tls_port`` - set this to any port you would like (but usually you
+would want 50052)
+
+``torii_tls_keypair`` - set this to full path to the key/certificate pair,
+so if you have a key at ``/path/to/server.key`` and a certificate at
+``/path/to/server.crt``, you need to specify
```
Comment (igor-egorov) [inline docs/source/guides/torii-tls.rst]:
```suggestion
``/path/to/server.crt``, you need to specify
```
trailing whitespace

File: irohad/main/application.hpp [hunk 6]
```diff
@@ -80,6 +80,8 @@ class Irohad {
    * @param pg_conn - initialization string for postgre
    * @param listen_ip - ip address for opening ports (internal & torii)
    * @param torii_port - port for torii binding
+   * @param torii_tls_port - port for torii TLS binding
+   * @param torii_tls_keypair - path to TLS keypair (.crt, .key) for torii
```
Comment (igor-egorov) [inline irohad/main/application.hpp]: Please add a remark about the path format. Should it include file extension or not?

Comment (igor-egorov) [inline irohad/main/application.hpp]: Please add a remark about the path format. Should it include file extension or not?

File: irohad/main/application.cpp [hunk 7]
```diff
@@ -741,6 +745,15 @@ Irohad::RunResult Irohad::run() {
       log_manager_->getChild("ToriiServerRunner")->getLogger(),
       false);
 
+  bool enable_tls = torii_tls_keypair_.length() > 0;
+  if (enable_tls) {
+    torii_tls_server = std::make_unique<ServerRunner>(
+        listen_ip_ + ":" + std::to_string(torii_tls_port_),
+        log_manager_->getChild("ToriiServerRunner")->getLogger(),
```
Comment (igor-egorov) [inline irohad/main/application.cpp]:
```suggestion
        log_manager_->getChild("ToriiTlsServerRunner")->getLogger(),
```

File: irohad/main/server_runner.hpp [hunk 8]
```diff
@@ -24,7 +24,8 @@ class ServerRunner {
    */
   explicit ServerRunner(const std::string &address,
                         logger::LoggerPtr log,
-                        bool reuse = true);
+                        bool reuse = true,
+                        const std::string &tls_keypair = "");
```
Comment (igor-egorov) [inline irohad/main/server_runner.hpp]: ``tls_keypair`` argument is not documented in the docstring above. Please add a comment.

File: irohad/main/server_runner.hpp [hunk 9]
```diff
@@ -64,6 +65,7 @@ class ServerRunner {
   std::string serverAddress_;
   bool reuse_;
   std::vector<std::shared_ptr<grpc::Service>> services_;
+  std::string tlsKeypair_;
```
Comment (igor-egorov) [inline irohad/main/server_runner.hpp]: We use ``camel_case_`` naming style for internal class members.
```suggestion
  std::string tls_keypair_;
```

File: irohad/main/server_runner.cpp [hunk 10]
```diff
@@ -5,15 +5,22 @@
 
 #include "main/server_runner.hpp"
 
+#include <fstream>
```
Comment (igor-egorov) [inline irohad/main/server_runner.cpp]: According to the Google c++ style guide any STL includes have to be divided from the others with a blank line.
Please add a blank line.

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: irohad/main/server_runner.cpp [hunk 11]
```diff
@@ -5,15 +5,22 @@
...
 ServerRunner::ServerRunner(const std::string &address,
                            logger::LoggerPtr log,
-                           bool reuse)
-    : log_(std::move(log)), serverAddress_(address), reuse_(reuse) {}
+                           bool reuse,
+                           const std::string &tls_keypair)
+    : log_(std::move(log)),
+      serverAddress_(address),
+      reuse_(reuse),
+      tlsKeypair_(tls_keypair) {}
```
Comment (igor-egorov) [inline irohad/main/server_runner.cpp]: tls_keypair_. serverAddress_ can be refactored too since you have been touched that part of the code :)

Comment (igor-egorov) [inline irohad/main/server_runner.cpp]: tls_keypair_. serverAddress_ can be refactored too since you have been touched that part of the code :)

File: irohad/main/server_runner.cpp [hunk 12]
```diff
@@ -28,8 +35,37 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
...
+    }
+
+    grpc::SslServerCredentialsOptions::PemKeyCertPair keypair = {
+      private_key_data, certificate_data
+    };
+    auto options = grpc::SslServerCredentialsOptions();
+    options.pem_key_cert_pairs.push_back(keypair);
+    auto credentials = grpc::SslServerCredentials(options);
+    builder.AddListeningPort(serverAddress_, credentials, &selected_port);
+  } else { // if (enable_tls)
```
Comment (igor-egorov) [inline irohad/main/server_runner.cpp]:
```suggestion
  } else { // tls is not enabled
```

File: irohad/main/server_runner.cpp [hunk 13]
```diff
@@ -28,8 +35,37 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
     builder.AddChannelArgument(GRPC_ARG_ALLOW_REUSEPORT, 0);
   }
 
-  builder.AddListeningPort(
-      serverAddress_, grpc::InsecureServerCredentials(), &selected_port);
+  bool enable_tls = tlsKeypair_.length() > 0;
+  if (enable_tls) {
```
Comment (igor-egorov) [inline irohad/main/server_runner.cpp]: Probably the whole TLS initialization code can be moved to a separate method.
Please consider that option. Otherwise, ``run`` method comes quite big.

File: scripts/run-iroha-dev.sh [hunk 14]
```diff
@@ -18,16 +18,19 @@ export COMPOSE_PROJECT_NAME=${PROJECT}
...
 then
     export IROHA_PORT="$(next_free_port 50051 50101)"
+    export IROHA_TLS_PORT="$(next_free_port 50052 50101)"
     export DEBUGGER_PORT="$(next_free_port 20000 20100)"
 
     docker-compose -f ${COMPOSE} up -d
 else
-    IROHA_DBG_PORTS="$(docker port ${PROJECT}_node_1 | sed 's/\(.*\)://' | sort -r | sed -e :a -e N -e 's/\n/:/p' -e ta)"
-    export IROHA_PORT="$(echo ${IROHA_DBG_PORTS} | sed 's/:.*//')"
+    IROHA_DBG_PORTS="$(docker port ${PROJECT}_node_1 | sed 's/\(.*\)://' | sort -r | paste -s -d ':' -)"
```
Comment (igor-egorov) [inline scripts/run-iroha-dev.sh]: 👀

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 15]
```diff
@@ -154,6 +156,8 @@ namespace integration_framework {
                                             block_store_path,
                                             kLocalHost,
                                             torii_port_,
+                                            0,
```
Comment (igor-egorov) [inline test/framework/integration_framework/integration_test_framework.cpp]: Why you are not using here the value of previously defined constant ``constexpr size_t kDefaultToriiTlsPort = 0;``?
I believe that will make the code more clear.

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 16]
```diff
@@ -154,6 +156,8 @@ namespace integration_framework {
...
+                                            "",
```
Comment (igor-egorov) [inline test/framework/integration_framework/integration_test_framework.cpp]: Not cool.
Probably we need some struct like this:
```
struct TlsConfig {
    size_t port,
    std::string key_path
};
```
Then, two initialization parameters can be substituted by the structure.
Moreover, looks like that combined argument can be declared as ``boost::optional`` because it is not a required initialization argument.

File: test/framework/integration_framework/iroha_instance.hpp [hunk 17]
```diff
@@ -47,6 +47,8 @@ namespace integration_framework {
                   const std::string &block_store_path,
                   const std::string &listen_ip,
                   size_t torii_port,
+                  size_t torii_tls_port,
```
Comment (igor-egorov) [inline test/framework/integration_framework/iroha_instance.hpp]: Look at my comment in `integration_test_framework.cpp` about possible `struct TlsConfig`;

Comment (igor-egorov) [reply test/framework/integration_framework/iroha_instance.hpp]: Newly introduced arguments are not documented in a docstring above.

File: docs/source/guides/torii-tls.rst [hunk 18]
```diff
@@ -0,0 +1,51 @@
+Configure TLS for client-peer communication (torii)
+===================================================
+By default, client-peer communication is not encrypted.
+To enable it, you need to:
+
+1. `Generate <Generating keys>`_ a key/certificate pair for each peer
```
[the remark this answers is not in the document]
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version, they did not compile correctly, but now do

File: irohad/main/server_runner.hpp (hunk 9 above)
Comment (ionagamed) [reply irohad/main/server_runner.hpp]: There is `serverAddress_` literally 4 lines before this one, so I was confused :)

File: irohad/main/server_runner.cpp (hunk 10 above)
Comment (ionagamed) [reply irohad/main/server_runner.cpp]: Fixed

File: docs/source/guides/torii-tls.rst [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+4. [Re]start irohad
+
+
+Generating keys
+~~~~~~~~~~~~~~~
+
+Keys must be presented in PEM format. To generate them you can use ``openssl``:
+
+```bash
+$ openssl genpkey -algorithm <desired algorithm> -out server.key
```
[the remark this answers is not in the document]
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version

File: docs/source/guides/torii-tls.rst [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+If you need to use plain IP addresses to connect to the node, you need to
+specify subjectAltName in your server certificate, for that you need to add
+a ``subjectAltName`` directive to ``v3_ca`` section of your openssl config 
+before generating the certificate. 
+For example, for the default installation, ``/etc/ssl/openssl.cnf``:
+
+```conf
+[ v3_ca ]
+subjectAltName=IP:12.34.56.78
+...
```
[the remark this answers is not in the document]
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version

File: docs/source/guides/torii-tls.rst (hunk 2 above)
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version

File: docs/source/guides/torii-tls.rst (hunk 3 above)
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version

File: docs/source/guides/torii-tls.rst (hunk 4 above)
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version

File: docs/source/guides/torii-tls.rst (hunk 5 above)
Comment (ionagamed) [reply docs/source/guides/torii-tls.rst]: Fixed in latest version

File: irohad/main/application.cpp (hunk 7 above)
Comment (ionagamed) [reply irohad/main/application.cpp]: Fixed in latest version

File: irohad/main/server_runner.hpp (hunk 9 above)
Comment (ionagamed) [reply irohad/main/server_runner.hpp]: Fixed in latest version

File: irohad/main/server_runner.hpp (hunk 8 above)
Comment (ionagamed) [reply irohad/main/server_runner.hpp]: Fixed in latest version

File: irohad/main/application.hpp (hunk 6 above)
Comment (ionagamed) [reply irohad/main/application.hpp]: Fixed in latest version

Comment (ionagamed) [reply irohad/main/application.hpp]: Fixed in latest version

File: irohad/main/server_runner.cpp (hunk 11 above)
Comment (ionagamed) [reply irohad/main/server_runner.cpp]: Fixed in latest version

Comment (ionagamed) [reply irohad/main/server_runner.cpp]: Fixed in latest version

File: irohad/main/server_runner.cpp (hunk 12 above)
Comment (ionagamed) [reply irohad/main/server_runner.cpp]: Fixed in latest version

File: irohad/main/server_runner.cpp (hunk 13 above)
Comment (ionagamed) [reply irohad/main/server_runner.cpp]: Fixed in latest version

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 15 above)
Comment (ionagamed) [reply test/framework/integration_framework/integration_test_framework.cpp]: Fixed in latest version

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 16 above)
Comment (ionagamed) [reply test/framework/integration_framework/integration_test_framework.cpp]: Fixed in latest version. This struct is available as `iroha::torii::TlsParams`

File: test/framework/integration_framework/iroha_instance.hpp (hunk 17 above)
Comment (ionagamed) [reply test/framework/integration_framework/iroha_instance.hpp]: Fixed in latest version

File: docker/docker-compose.yml [hunk 21]
```diff
@@ -5,6 +5,7 @@ services:
     image: hyperledger/iroha:develop-build
     ports:
       - "${IROHA_PORT}:50051"
+      - "${IROHA_TLS_PORT}:50052"
```
[the remark this answers is not in the document]
Comment (ionagamed) [reply docker/docker-compose.yml]: Fixed in latest version

[the remark this answers is not in the document]
Comment (ionagamed) [reply docker/docker-compose.yml]: Fixed in latest version. Port was reassigned to 55552

**Comment by ionagamed** [comment]
Fixed all the comments. This is ready for a re-review

File: irohad/main/iroha_conf_literals.hpp [hunk 22]
```diff
@@ -14,6 +14,8 @@
 namespace config_members {
   extern const char *BlockStorePath;
   extern const char *ToriiPort;
+  extern const char *ToriiTlsPort;
+  extern const char *ToriiTlsKeypair;
```
Comment (MBoldyrev) [inline irohad/main/iroha_conf_literals.hpp]: add
```c++
extern const char *ToriiTlsParams
```

File: irohad/main/iroha_conf_loader.cpp [hunk 23]
```diff
@@ -355,6 +355,9 @@ inline void JsonDeserializerImpl::getVal<IrohadConfig>(
   const auto obj = src.GetObject();
   getValByKey(path, dest.block_store_path, obj, config_members::BlockStorePath);
   getValByKey(path, dest.torii_port, obj, config_members::ToriiPort);
+  getValByKey(path, dest.torii_tls_port, obj, config_members::ToriiTlsPort);
```
Comment (MBoldyrev) [inline irohad/main/iroha_conf_loader.cpp]:
```suggestion
  getValByKey(path, dest.torii_tls_params, obj, config_members::ToriiTlsParams);
```

**Comment by ionagamed** [comment]
Fixed all previous comments from @MBoldyrev, now the config can have

```json
{
    "torii_tls_params": {
        "port": 55552,
        "key_pair_path": "/path/to/a/keypair"
    }
}
```

**Comment by igor-egorov** [comment]
need to update branch

**Comment by ionagamed** [comment]
Rebased on top of target branch

File: irohad/main/server_runner.hpp [hunk 24]
```diff
@@ -44,6 +48,23 @@ class ServerRunner {
    */
   void waitForServersReady();
 
+  /**
+   * Construct SSL credentials for a secure connection
+   * @return Result with server credentials or error message
+   */
+  iroha::expected::Result<std::shared_ptr<grpc::ServerCredentials>, std::string>
+  createSecureCredentials();
```
Comment (lebdron) [inline irohad/main/server_runner.hpp]: There methods should definitely not be public. Please make them `private` or `protected`, or move them to an anonymous namespace in cpp file.

File: test/system/irohad_test.cpp [hunk 25]
```diff
@@ -157,6 +158,42 @@ class IrohadTest : public AcceptanceFixture {
...
+    }
+
+    std::unique_ptr<iroha::protocol::CommandService_v1::Stub> stub;
+    std::string stub_error;
+    if (enable_tls) {
+      iroha::network::createSecureClient<iroha::protocol::CommandService_v1>(
+          kAddress + ":" + std::to_string(port),
+          (test_data_path_ / "tls/correct.crt").string())
+          .match([&](auto secure_stub) { stub = std::move(secure_stub.value); },
+                 [&](const auto &error) { stub_error = error.error; });
```
Comment (lebdron) [inline test/system/irohad_test.cpp]: Maybe print `stub_error` somewhere or return it?

File: irohad/main/server_runner.hpp [hunk 26]
```diff
@@ -57,13 +78,14 @@ class ServerRunner {
  private:
   logger::LoggerPtr log_;
 
-  std::unique_ptr<grpc::Server> serverInstance_;
-  std::mutex waitForServer_;
-  std::condition_variable serverInstanceCV_;
+  std::unique_ptr<grpc::Server> server_instance_;
+  std::mutex wait_for_server_;
+  std::condition_variable server_instance_cv_;
```
Comment (MBoldyrev) [inline irohad/main/server_runner.hpp]: :snake: :+1:

File: test/system/irohad_test.cpp [hunk 27]
```diff
@@ -157,6 +158,42 @@ class IrohadTest : public AcceptanceFixture {
...
+    } else {
+      stub = iroha::network::createClient<iroha::protocol::CommandService_v1>(
+          kAddress + ":" + std::to_string(port));
+    }
+
+    return torii::CommandSyncClient(
+        std::move(stub),
```
Comment (lebdron) [inline test/system/irohad_test.cpp]: It does not look like it will work if there was an error in `createSecureClient`.

File: test/system/irohad_test.cpp [hunk 28]
```diff
@@ -297,6 +333,48 @@ TEST_F(IrohadTest, SendTx) {
   sendDefaultTxAndCheck(key_pair.get());
 }
 
+/**
+ * Test verifies that a transaction can be sent to running iroha and commited,
+ * through a TLS port
+ * @given running Iroha with an open TLS port
+ * @when a client sends a transaction to Iroha AND the server's certificate
+ *       is valid
+ * @then the transaction is commited
```
Comment (lebdron) [inline test/system/irohad_test.cpp]:
```suggestion
 * @then the transaction is committed
```
same below

File: test/system/irohad_test.cpp [hunk 29]
```diff
@@ -157,6 +158,42 @@ class IrohadTest : public AcceptanceFixture {
...
+      const boost::optional<uint16_t> override_port = {}) {
+    uint16_t port;
+    if (override_port) {
+      port = *override_port;
+    } else {
+      if (enable_tls) {
+        port = kSecurePort;
+      } else {
+        port = kPort;
+      }
```
Comment (MBoldyrev) [inline test/system/irohad_test.cpp]:
```c++
    uint16_t port = override_port.value_or(enable_tls ? kSecurePort : kPort);
```
instead of 9 lines above

File: test/system/irohad_test_data/config.sample [hunk 30]
```diff
@@ -1,8 +1,12 @@
 {
   "block_store_path" : "/tmp/block_store/",
   "torii_port" : 50051,
+  "torii_tls_params": {
+    "port": 55552,
+    "key_pair_path": "/opt/iroha/test/system/irohad_test_data/tls/correct"
+  },
   "internal_port" : 10001,
-  "pg_opt" : "host=localhost port=5432 user=postgres password=mysecretpassword",
+  "pg_opt" : "host=localhost port=15432 user=iroha password=helloworld",
```
Comment (MBoldyrev) [inline test/system/irohad_test_data/config.sample]: why is this change needed?

**Comment by lebdron** [comment]
Please also check `stop-iroha-dev.sh`

**Comment by ionagamed** [comment]
Fixed all the comments, except the one about adding another port to the `ServerRunner`.

Now the arguments are passed through the constructor, and then everything happens in the `run()` method. There are two ways I see that multiple ports could be implemented in `ServerRunner`:
1. Pass it through the constructor, via an optional container for both port and 
    certificate (e.g. `iroha::torii::TlsParams`).
2. Expose the interface to add as many listening ports, e.g.
    ```cpp
    iroha::expected::Result<...> addListeningPort(uint16_t, boost::optional<TLSKeypair>)
    ```
And I'm not sure which one to implement.

Otherwise, this is ready for a re-review.

File: example/config.docker [hunk 31]
```diff
@@ -2,7 +2,7 @@
   "block_store_path" : "/tmp/block_store/",
   "torii_port" : 50051,
   "internal_port" : 10001,
-  "pg_opt" : "host=some-postgres port=5432 user=postgres password=mysecretpassword",
+  "pg_opt" : host=some-postgres port=5432 user=postgres password=mysecretpassword",
```
Comment (MBoldyrev) [inline example/config.docker]: whoops

File: irohad/main/CMakeLists.txt [hunk 32]
```diff
@@ -5,7 +5,7 @@
 
 set(EXECUTABLE_OUTPUT_PATH ${PROJECT_BINARY_DIR}/bin)
 
-add_library(server_runner server_runner.cpp)
+add_library(server_runner server_runner.cpp tls_keypair.cpp)
```
Comment (MBoldyrev) [inline irohad/main/CMakeLists.txt]: please follow the multiline style for targets with multiple source files

File: irohad/main/application.hpp [hunk 33]
```diff
@@ -14,9 +14,11 @@
 #include "interfaces/queries/query.hpp"
 #include "logger/logger_fwd.hpp"
 #include "logger/logger_manager_fwd.hpp"
+#include "main/server_runner.hpp"
```
Comment (MBoldyrev) [inline irohad/main/application.hpp]: why did you change the forward declaration for header inclusion?

File: irohad/main/server_runner.cpp [hunk 34]
```diff
@@ -39,36 +47,68 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
...
-  while (not serverInstance_) {
-    serverInstanceCV_.wait(lock);
+  std::unique_lock<std::mutex> lock(wait_for_server_);
+  while (not server_instance_) {
+    server_instance_cv_.wait(lock);
+  }
+}
+
+iroha::expected::Result<std::shared_ptr<grpc::ServerCredentials>, std::string>
+ServerRunner::createSecureCredentials() {
```
Comment (MBoldyrev) [inline irohad/main/server_runner.cpp]: is this function supposed to ever return *error*? In its body i can see no such option. If so, why wrapping returned object into `Result`?

File: irohad/main/server_runner.cpp [hunk 35]
```diff
@@ -39,36 +47,68 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
...
+    createSecureCredentials().match(
+        [&](const auto &secure_credentials) {
+          credentials = secure_credentials.value;
+        },
+        [&](const auto &error) { credentials_error = error.error; });
+    if (credentials_error) {
+      return credentials_error;
+    }
+
+    builder.AddListeningPort(server_address_, credentials, selected_port);
```
Comment (MBoldyrev) [inline irohad/main/server_runner.cpp]: How are the keys having incorrect format handled? I am afraid, an exception may arise on this line or somewhere later. If this is the case, should catch such an exception and provide the user with an appropriate error message.
Also please check what happens if we provide a syntactically correct key that does not match the certificate.

File: test/framework/integration_framework/iroha_instance.hpp [hunk 36]
```diff
@@ -17,6 +17,7 @@
 #include "logger/logger_fwd.hpp"
 #include "logger/logger_manager_fwd.hpp"
 #include "multi_sig_transactions/gossip_propagation_strategy_params.hpp"
+#include "torii/tls_params.hpp"
```
Comment (MBoldyrev) [inline test/framework/integration_framework/iroha_instance.hpp]: Types wrapped into boost::optional suffice with a forward declaration.

**Comment by MBoldyrev** [comment]
Thanks!

> I'm not sure which one to implement.

My vote is for the second, although it could be done in a separate PR as a refactoring task.

File: irohad/main/application.hpp (hunk 33 above)
Comment (ionagamed) [reply irohad/main/application.hpp]: `boost::optional<std::shared_ptr<ServerRunner>>` seems to require a full type, otherwise my compiler gives `error: invalid application of 'sizeof' to an incomplete type 'ServerRunner'` (clang-10.0.1 on macos)

Comment (MBoldyrev) [reply irohad/main/application.hpp]: Tested it. Seems you are right, and I am not. Then this include is OK.

File: test/framework/integration_framework/iroha_instance.hpp (hunk 36 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/iroha_instance.hpp]: Same as above, my fault.

File: irohad/main/server_runner.cpp (hunk 35 above)
Comment (ionagamed) [reply irohad/main/server_runner.cpp]: I could not yet find how to handle this error getting a cause from gRPC, but invalid formats and mismatched certificates would both give `selected_port == 0`, and would print an error message, such as `E0714 18:22:43.053279000 4790195648 ssl_transport_security.cc:675]     Invalid cert chain file.` What would you suggest, do I keep digging (I have literally spent the weekend looking for a solution), or this would be sufficient?

Comment (MBoldyrev) [reply irohad/main/server_runner.cpp]: I think that if you checked various error cases and irohad 1) points up the correct error reason and 2) terminates gracefully - that is enough for now. Thank you for devoting your time to the investigation!

**Comment by ionagamed** [comment]
Fixed all previous comments (missing quote, redundant `Result` and `CMakeLists.txt` style). Probably ready for a re-review (I hope :) )

File: irohad/main/tls_keypair.cpp [hunk 37]
```diff
@@ -0,0 +1,22 @@
...
+  std::string readFile(const std::string &path) {
+    std::ifstream file(path);
+    std::stringstream ss;
+    ss << file.rdbuf();
+    return ss.str();
+  }
+}
+
+TLSKeypair::TLSKeypair(const std::string &path)
+    : certificate(readFile(path + ".crt")),
```
Comment (lebdron) [inline irohad/main/tls_keypair.cpp]: Basically these two `readFile` calls can be done somewhere in irohad.cpp, then two `Blob` objects are initialized from `readFile` result, and `TLSKeypair` is initialized from two `Blob` objects. This will allow to initialize `ServerRunner` not only from files.

File: irohad/main/server_runner.hpp [hunk 38]
```diff
@@ -21,10 +22,13 @@ class ServerRunner {
...
+   * @param tls_keypair - TLS keypair, if TLS is requested
    */
-  explicit ServerRunner(const std::string &address,
-                        logger::LoggerPtr log,
-                        bool reuse = true);
+  explicit ServerRunner(
+      const std::string &address,
+      logger::LoggerPtr log,
+      bool reuse = true,
+      const boost::optional<TLSKeypair> &tls_keypair = boost::none);
```
[the remark this answers is not in the document]
Comment (ionagamed) [reply irohad/main/server_runner.hpp]: `TLSKeypair` takes on the responsibility of reading the files, so I can maybe add a separate constructor to create a keypair from raw data?

File: irohad/main/tls_keypair.cpp (hunk 37 above)
Comment (ionagamed) [reply irohad/main/tls_keypair.cpp]: Oh, okay, nevermind the last comment, this seems more reasonable, will do so

**Comment by ionagamed** [comment]
Fixed the docs (it is now a valid json structure), and moved the file-reading logic into a separate `TLSKeypairFactory`. I didn't see much use in going with a `Blob` instead of `std::string`, because PEM-encoded keys are printable-text-only anyway, and gRPC requires the keys in PEM form

File: example/torii_tls/server.crt [hunk 39]
```diff
@@ -0,0 +1,16 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (lebdron) [reply example/torii_tls/server.crt]: Bump

File: irohad/main/tls_keypair.hpp [hunk 40]
```diff
@@ -8,17 +8,32 @@
...
+   *               appended.
+   *               @see iroha::torii::TlsParams
+   * @return keypair read from this path
+   */
+  boost::optional<TLSKeypair> readFromFiles(const std::string &path);
 };
 
-#endif // IROHA_TLS_KEYPAIR_HPP
\ No newline at end of file
+#endif  // IROHA_TLS_KEYPAIR_HPP
```
Comment (lebdron) [inline irohad/main/tls_keypair.hpp]: A newline would be great.
